### PR TITLE
New features for upcoming 1.2

### PR DESCRIPTION
**Pull request by dchirikov**
* DHCP IP reservation for nodes
* Group clone operation
* Arbitrary kernel version for osimage
* Comments for all luna objects 
* DB migration script from 1.1

### CODE REVIEW

File: luna/osimage.py [hunk 1]
```diff
@@ -350,21 +354,33 @@ def cleanup_mounts(path):
...
         os.chroot(".")
         os.close(real_root)
         cleanup_mounts(image_path)
 
-        shutil.copy(image_path + tmp_path + '/' + initrdfile, path_to_store)
-        shutil.copy(image_path + '/boot/vmlinuz-' + kernver,
-                    path_to_store + '/' + kernfile)
+        if not dracut_succeed:
+            self.log.error("Error while building initrd.")
+            return False
```
Comment (hicham-a) [inline luna/osimage.py]: I think you forgot to add the cleanup code here (chroot, mounts)

File: luna/osimage.py [hunk 2]
```diff
@@ -350,21 +354,33 @@ def cleanup_mounts(path):
...
+        if not os.path.isfile(kernel_path):
+            self.log.error("Unable to find kernel in {}".format(kernel_path))
+            return False
+
+        if not os.path.isfile(initrd_path):
+            self.log.error("Unable to find initrd in {}".format(initrd_path))
+            return False
+
+        shutil.copy(initrd_path, path_to_store)
+        shutil.copy(kernel_path, path_to_store + '/' + kernfile)
```
Comment (hicham-a) [inline luna/osimage.py]: We can use shutil.move for the initrd. It'll make the image smaller by around 50MB

File: tests/test_group.py [hunk 3]
```diff
@@ -780,16 +782,17 @@ def setUp(self):
 
         # mocking osimage boot stuff
         self.osimage.copy_boot()
-        self.osimage.create_tarball()
-        self.osimage.create_torrent()
+        #self.osimage.create_tarball()
```
Comment (hicham-a) [inline tests/test_group.py]: These need to remain enabled I think until even though a fix isn't ready for them yet

File: bin/luna [hunk 4]
```diff
@@ -955,6 +955,7 @@ def group_change(**args):
 
     return True
 
+
 def group_clone(**args):
```
Comment (hicham-a) [inline bin/luna]: I think this logic should be in the Group class instead of the CLI.
Something like group.duplicate()

File: bin/luna [hunk 5]
```diff
@@ -79,6 +79,26 @@ def print_table(header, content):
...
+    script = ''
+    if not sys.stdin.isatty():
+        for line in sys.stdin:
+            script += line
+    else:
+        EDITOR = os.environ.get('EDITOR', 'vi')
+        tmpfile = tempfile.NamedTemporaryFile(suffix=".tmp")
+        tmpfile.write(old_script)
+        tmpfile.flush()
+        subprocess.call([EDITOR, '+":set syn=sh"', tmpfile.name])
```
Comment (dchirikov) [inline bin/luna]: Removed in ff10f34

File: bin/luna [hunk 6]
```diff
@@ -1032,6 +1134,11 @@ def node_change(**args):
     for nodename in hostlist.expand_hostlist(args['name']):
         node = luna.Node(nodename)
 
+        if args['comment']:
+            old_comment = node.get('comment')
+            # change only first node
+            return node.set('comment', _edit_script(old_comment))
```
Comment (dchirikov) [inline bin/luna]: Fixed in e3735c3

File: luna/group.py [hunk 7]
```diff
@@ -576,6 +581,84 @@ def get_allocated_ips(self, net):
...
+        if not self.get(usedby_key) or not 'node' in self.get(usedby_key):
+            self.log.warning(
+                'No nodes are configured in provisioning network.')
+            # No nodes in group. Returning empty list
+            return {}
+
+        # get all the interfaces with desired net configured in format:
+        # {'uuid1': 'eth0', 'uuid2': 'BMC'}
+        interfaces_with_net = {}
+        for ifuuid in interfaces:
```
Comment (dchirikov) [inline luna/group.py]: As disussed internally we will end up with mess considering corner cases as bonded interfaces.
For now it s decided to leave it as is and add 9a6f3c2

File: luna/osimage.py (hunk 2 above)
Comment (dchirikov) [inline luna/osimage.py]: Idea was to keep  it available in tmp in the case someone will need to run lsinitrd on the running node

**Comment by kostty** [comment]
looks fine to me

File: bin/luna [hunk 8]
```diff
@@ -838,89 +912,112 @@ def group_add(**args):
...
         return None
-    #if args['interface'] == 'BMC':
-    #    args['interface'] = None
-    #    args['bmcnetwork'] = True
     group = luna.Group(args['name'])
+    for txt_field in ['prescript', 'postscript',
+                      'partscript', 'comment']:
+        if args[txt_field]:
+            old_comment = group.get(txt_field)
+            return group.set(txt_field, _edit_script(old_comment))
```
Comment (dchirikov) [inline bin/luna]: Reasonable. I missed this part. Thought this comment is about other lines.
Changed in 6f3b462

File: luna/group.py [hunk 9]
```diff
@@ -576,6 +581,84 @@ def get_allocated_ips(self, net):
 
         return ips
 
+    def get_macs(self, net):
+
+        interfaces = self.get('interfaces')
+
+        if type(net) != Network:
```
Comment (dchirikov) [inline luna/group.py]: Valid point. Changed in 0c719d9

**Pull request merged**